# Detection of Malicicous Data

## Part I: Preprocessing 

### Read the data.file
The data file includes dataset structured like Basic Safety Messages (BSM) defined in the SEA J2735 standard, including vehicle's location, speed, acceleration, heading, and brake status. We utilize the data to check the correcteness of the data and better classify the five types of position spoofing attack. 

In [1]:
import pandas as pd 
import matplotlib.pyplot as plt 
from mpl_toolkits.mplot3d import Axes3D
import numpy as np

### colunm data
1. Type (3=BSM)
2. Time BSM was received by the receiver
3. Receiver ID
4. Receiver X position
5. Receiver Y position
6. Receiver Z position
7. Time BSM was transmitted
8. Transmitter ID
9. BSM ID
10. Transmitter X position
11. Transmitter Y position
12. Transmitter Z position
13. Transmitter X velocity
14. Transmitter Y velocity
15. Transmitter Z velocity
16. RSSI (Received Signal Strength Indicator)
17. Label (0=Normal Behavior)

In [20]:
cols = [2,3,4,9,10,15,16] #id, xr ,yr, xt, yt, RSSI, label.
data = pd.read_csv("dataset/attack1withlabels.csv",usecols=cols)
data = data.dropna(axis=0, how="any")#remove invalid data
data_id = data.iloc[:,0]
# attack 1: distance is too far recevier and transmitter
pos_xr = data.iloc[:,1]
pos_yr = data.iloc[:,2]
pos_xt = data.iloc[:,3]
pos_yt = data.iloc[:,4]
rssi   = data.iloc[:,5]
label = data.iloc[:,6]


In [23]:
# We read the data from the dataset.
data = pd.read_csv("dataset/attack1withlabels.csv", usecols=cols, header=None)
X = data.iloc[:,:6]
y = data.iloc[:,6]

Two observations for attack 1
* The fake distance exceed the upper bound of communication range which is 800m
* The distance does not change but the RSSI changes.
The first criteria is trivial, thus we apply a filter. 

In [26]:
#add a distance feature
def distance(X):
    distance = np.zeros(len(X))
    for i in range(len(X)):
        distance[i] = np.linalg.norm([X.iloc[i][3]-X.iloc[i][9], X.iloc[i][4]-X.iloc[i][10]])
    distance_series = pd.Series(distance)
    return distance_series

distance_series = distance(X)
X['distance'] = distance_series

In [25]:
# We filter the BSM whose distance is above the thredhold = 800
def filtering_dis(X,y,thredhold):
    drop_index = np.zeros(len(y))
    for i in range(len(y)):
        if X['distance'][i] > thredhold:
            drop_index[i]=1
    return drop_index
drop_index = filtering_dis(X,y,800)
drop_index_ = np.where(drop_index>0)
np.asarray(drop_index_)[0]
X_filter = X.drop(np.asarray(drop_index_)[0])
y_filter = y.drop(np.asarray(drop_index_)[0])

2       3       4       9       10            15
0        562  5607.3  5965.1  5367.5  5930.1  1.420000e-08
1        562  5594.2  5973.5  5373.5  5934.6  3.120000e-09
2        562  5583.8  5980.2  5368.1  5924.2  2.260000e-09
3       3166  3642.0  5183.0  5560.0  5820.0  3.790000e-09
4       3166  3642.0  5183.0  3781.6  5256.0  3.150000e-09
5       3166  3642.0  5183.0  5560.0  5820.0  5.580000e-09
6       3166  3642.0  5183.0  4168.2  5286.1  2.940000e-09
7       3166  3642.0  5183.0  3609.1  5420.4  2.070000e-08
8       3166  3641.4  5186.5  3485.7  5192.2  3.580000e-09
9       3166  3640.9  5190.1  5560.0  5820.0  1.580000e-09
10      3166  3640.9  5190.1  3596.1  5514.8  1.240000e-08
11      3166  3640.9  5190.1  3479.0  5190.9  4.180000e-09
12      3166  3640.9  5190.1  3600.3  5503.8  2.350000e-09
13      3166  3640.9  5190.1  3617.9  5373.6  9.110000e-09
14      3166  3639.8  5197.1  3632.4  5348.3  3.770000e-09
15      3166  3639.2  5200.7  3634.0  5310.3  6.640000e-08
16      3166  3639.2  5200.7  3479.6  5187.7  5.860000e-09
17      3166  3639.2  5200.7  3962.5  5245.2  1.300000e-09
18      3166  3639.2  5200.7  5560.0  5820.0  1.090000e-08
19      3166  3638.1  5207.7  3636.5  5273.4  3.230000e-08
20      3166  3638.1  5207.7  3795.2  5269.5  6.940000e-09
21      3166  3638.1  5207.7  3642.2  5202.6  1.110000e-06
22      3166  3638.1  5207.7  5560.0  5820.0  5.460000e-09
23      3166  3638.1  5207.7  3627.0  5257.8  2.430000e-07
24      3166  3637.6  5211.2  3606.1  5418.1  1.900000e-08
25      3166  3637.6  5211.2  5560.0  5820.0  4.210000e-08
26      3166  3637.6  5211.2  5560.0  5820.0  8.380000e-08
27      3166  3637.6  5211.2  3949.1  5256.8  1.040000e-08
28      3166  3637.6  5211.2  5560.0  5820.0  5.240000e-08
29      3166  3637.0  5214.8  3480.9  5184.6  5.770000e-09
...      ...     ...     ...     ...     ...           ...
425678  1486  3586.8  5981.6  3604.5  6002.7  4.070000e-08
425679  1486  3586.8  5981.6  3591.2  5777.5  3.850000e-09
425680  1486  3587.1  5985.5  3600.3  5627.6  5.310000e-09
425681  1486  3587.1  5985.5  3601.5  5660.4  1.330000e-08
425682  1486  3587.4  5989.3  3595.1  6068.6  2.460000e-08
425683  1486  3587.4  5989.3  3597.2  5682.6  2.260000e-09
425684  1486  3587.9  5997.1  3602.5  5517.5  1.880000e-09
425685  1486  3587.9  5997.1  3605.7  5479.1  1.790000e-09
425686  1486  3588.2  6000.9  5560.0  5820.0  6.230000e-09
425687  1486  3588.2  6000.9  3600.4  5685.7  4.170000e-09
425688  1486  3588.2  6000.9  5560.0  5820.0  5.910000e-09
425689  1486  3588.2  6000.9  3622.3  5375.7  1.690000e-09
425690  1486  3588.5  6004.8  3596.0  5710.4  4.390000e-09
425691  1486  3588.8  6008.6  3595.9  5918.1  1.040000e-07
425692  1486  3590.0  6020.2  3601.9  5964.5  1.600000e-07
425693  1486  3590.0  6020.2  3591.3  5744.0  3.140000e-09
425694  1486  3590.4  6024.1  3602.9  5595.9  6.400000e-09
425695  1486  3590.8  6027.9  3598.7  5648.9  7.720000e-09
425696  1486  3591.6  6035.6  3603.1  5510.4  2.690000e-09
425697  1486  3592.1  6039.5  5560.0  5820.0  9.600000e-09
425698  1486  3592.1  6039.5  5560.0  5820.0  5.250000e-09
425699  1486  3592.5  6043.3  3597.5  5676.6  7.580000e-09
425700  1486  3592.9  6047.2  3594.5  5881.6  7.830000e-09
425701  1486  3594.1  6058.7  3599.5  5926.5  3.850000e-09
425702  1486  3594.1  6058.7  3592.7  5710.5  6.130000e-09
425703  1486  3594.5  6062.6  3605.6  5564.1  2.420000e-09
425704  1486  3594.5  6062.6  3606.1  5597.4  4.610000e-09
425705  1486  3594.5  6062.6  3625.3  5405.5  1.400000e-09
425706  1486  3595.7  6074.1  3597.3  5580.3  3.670000e-09
425707  1486  3596.1  6078.0  5560.0  5820.0  2.750000e-09

[425708 rows x 6 columns]

In [46]:
#drop receiver colums 
X_feed = X_filter.drop([3, 4], axis = 1)
X_feed

2       9      10            15    distance
0        562  5367.5  5930.1  1.420000e-08  242.340752
1        562  5373.5  5934.6  3.120000e-09  224.101986
2        562  5368.1  5924.2  2.260000e-09  222.850825
4       3166  3781.6  5256.0  3.150000e-09  157.534631
6       3166  4168.2  5286.1  2.940000e-09  536.205231
7       3166  3609.1  5420.4  2.070000e-08  239.668876
8       3166  3485.7  5192.2  3.580000e-09  155.804300
10      3166  3596.1  5514.8  1.240000e-08  327.776036
11      3166  3479.0  5190.9  4.180000e-09  161.901977
12      3166  3600.3  5503.8  2.350000e-09  316.316376
13      3166  3617.9  5373.6  9.110000e-09  184.935800
14      3166  3632.4  5348.3  3.770000e-09  151.380976
15      3166  3634.0  5310.3  6.640000e-08  109.723288
16      3166  3479.6  5187.7  5.860000e-09  160.128573
17      3166  3962.5  5245.2  1.300000e-09  326.348188
19      3166  3636.5  5273.4  3.230000e-08   65.719480
20      3166  3795.2  5269.5  6.940000e-09  168.818394
21      3166  3642.2  5202.6  1.110000e-06    6.543699
23      3166  3627.0  5257.8  2.430000e-07   51.314910
24      3166  3606.1  5418.1  1.900000e-08  209.284161
27      3166  3949.1  5256.8  1.040000e-08  314.819964
29      3166  3480.9  5184.6  5.770000e-09  158.994497
30      3166  3643.9  5226.2  1.950000e-06   13.325539
31      3166  3817.1  5262.9  5.850000e-09  186.412500
34      3166  3485.4  5192.2  2.730000e-09  153.383213
35      3166  3592.8  5555.1  4.130000e-09  332.440747
36      3166  3479.0  5190.9  5.360000e-09  160.159951
37      3166  3597.3  5539.9  1.330000e-08  316.799400
38      3166  3614.4  5403.0  1.270000e-08  178.837244
40      3166  3589.9  5698.2  2.470000e-09  471.442998
...      ...     ...     ...           ...         ...
425673  1486  3599.0  5717.3  7.420000e-09  245.377179
425674  1486  3618.5  5406.6  1.780000e-09  556.684920
425675  1486  3594.6  5744.2  1.260000e-08  222.182358
425676  1486  3598.0  5954.6  3.370000e-07   19.584943
425677  1486  3598.8  6071.9  4.830000e-08  102.713290
425678  1486  3604.5  6002.7  4.070000e-08   27.540879
425679  1486  3591.2  5777.5  3.850000e-09  204.147422
425680  1486  3600.3  5627.6  5.310000e-09  358.143337
425681  1486  3601.5  5660.4  1.330000e-08  325.418761
425682  1486  3595.1  6068.6  2.460000e-08   79.672957
425683  1486  3597.2  5682.6  2.260000e-09  306.856530
425684  1486  3602.5  5517.5  1.880000e-09  479.822175
425685  1486  3605.7  5479.1  1.790000e-09  518.305740
425687  1486  3600.4  5685.7  4.170000e-09  315.436016
425689  1486  3622.3  5375.7  1.690000e-09  626.129260
425690  1486  3596.0  5710.4  4.390000e-09  294.495518
425691  1486  3595.9  5918.1  1.040000e-07   90.778081
425692  1486  3601.9  5964.5  1.600000e-07   56.957001
425693  1486  3591.3  5744.0  3.140000e-09  276.203059
425694  1486  3602.9  5595.9  6.400000e-09  428.382411
425695  1486  3598.7  5648.9  7.720000e-09  379.082326
425696  1486  3603.1  5510.4  2.690000e-09  525.325889
425699  1486  3597.5  5676.6  7.580000e-09  366.734086
425700  1486  3594.5  5881.6  7.830000e-09  165.607729
425701  1486  3599.5  5926.5  3.850000e-09  132.310241
425702  1486  3592.7  5710.5  6.130000e-09  348.202814
425703  1486  3605.6  5564.1  2.420000e-09  498.623565
425704  1486  3606.1  5597.4  4.610000e-09  465.344603
425705  1486  3625.3  5405.5  1.400000e-09  657.821442
425706  1486  3597.3  5580.3  3.670000e-09  493.802592

[333003 rows x 5 columns]

In [72]:
# split training data and validate data
data_count = y_filter.count()
data_rate = 0.8 * data_count
data_rate = int(data_rate)
X_train = X_feed.iloc[:data_rate,:]
y_train = y_filter.iloc[:data_rate]
X_test = X_feed.iloc[data_rate:,:]
y_test = y_filter.iloc[data_rate:]

X_train.shape

(266402, 5)

After pre-processing, we get the following relevant information:
1. The ID of the transmitter BSM (Coloum 2)
2. Transmitter positions (Coloum 9,10)
3. The distance between transmitter and receiver (Coloum 'distance')
4. The RSSI (Coloum 16)
We feed the following information into the training model.

## impletation keras

In [73]:
from tensorflow.python.keras import backend as k
from keras.models import Sequential
from keras.layers import Dense,Activation
from keras.layers.recurrent import LSTM
import tensorflow
from sklearn import model_selection


In [74]:
# model implementation
model = Sequential()
model.add(LSTM(32, input_shape=(None,5),return_sequences=False))
model.add(Dense(8, input_dim=5))
model.add(Dense(1, activation="relu"))
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [79]:
print("The radio of 0/all = ",1-sum(y_filter)/y_filter.shape[0])
X_train.shape

The radio of 0/all =  0.8671063023456245


(266402, 1, 5)

In our dataset, attack take the part of 13.3% and normal take a part of 86.7% 

In [ ]:
#to satisfait LSTM
X_train = np.reshape(X_train.values, (X_train.shape[0], 1,X_train.shape[1]))
X_test =np.reshape(X_test.values, (X_test.shape[0], 1, X_test.shape[1]))

In [ ]:
#fit data
model.fit(X_train,y_train,epochs=10,batch_size=100)

In [125]:
cost = model.evaluate(X_test.values,y_test.values,batch_size=100)
print('loss \n',cost[0])
print('accurency \n',cost[1])

66601/66601 [==============================] - 1s 8us/step
loss 
 1.8329824162138717
accurency 
 0.8862779837549901
